be/src/exprs/agg/maxmin_by.h

```diff
@@ -0,0 +1,329 @@
+    DateValue max = DateValue::MIN_DATE_VALUE;
+    void reset() {
+        buffer_result.clear();
+        max = DateValue::MIN_DATE_VALUE;
+    }
+};
+
+
+template <PrimitiveType PT, typename State, typename = guard::Guard>
+struct MaxByElement {
+    using T = RunTimeCppType<PT>;
+    void operator()(State& state, Column* col, size_t row_num, const T& right) const {
+        if (right > state.max) {
+            state.max = right;
+            state.buffer_result.resize(col->serialize_size(row_num));
+            col->serialize(row_num, state.buffer_result.data());
+        }
+    }
+    void operator()(State& state, const char* buffer, size_t size, const T& right) const {
+        if (right > state.max) {
+            state.max = right;
+            state.buffer_result.resize(size);
+            memcpy(state.buffer_result.data(), buffer, size);
+        }
+    }
+};
+
+template <PrimitiveType PT>
+struct MaxByAggregateData<PT, BinaryPTGuard<PT>> {
+    raw::RawVector<uint8_t> buffer_result;
+    raw::RawVector<uint8_t> buffer_max;
+    Slice slice_max() const { return {buffer_max.data(), buffer_max.size()}; }
+    void reset() {
+        buffer_result.clear();
+        buffer_max.clear();
+    }
+};
+
+template <PrimitiveType PT>
+struct MaxByElement<PT, MaxByAggregateData<PT>, BinaryPTGuard<PT>> {
+    void operator()(MaxByAggregateData<PT>& state, Column* col, size_t row_num, const Slice& right) const {
+        if (state.buffer_max.empty() || state.slice_max().compare(right) < 0) {
+            state.buffer_result.resize(col->serialize_size(row_num));
+            col->serialize(row_num, state.buffer_result.data());
+            state.buffer_max.resize(right.size);
+            memcpy(state.buffer_max.data(), right.data, right.size);
+        }
+    }
+
+    void operator()(MaxByAggregateData<PT>& state, const char* buffer, size_t size, const Slice& right) const {
+        if (state.buffer_max.empty() || state.slice_max().compare(right) < 0) {
+            state.buffer_result.resize(size);
+            memcpy(state.buffer_result.data(), buffer, size);
+            state.buffer_max.resize(right.size);
+            memcpy(state.buffer_max.data(), right.data, right.size);
+        }
+    }
+};
+
+
+template <PrimitiveType PT, typename State, class OP, typename T = RunTimeCppType<PT>, typename = guard::Guard>
+class MaxByAggregateFunction final
+        : public AggregateFunctionBatchHelper<State, MaxByAggregateFunction<PT, State, OP, T>> {
+public:
+    using InputColumnType = RunTimeColumnType<PT>;
+
+    void reset(FunctionContext* ctx, const Columns& args, AggDataPtr state) const override {
+        this->data(state).reset();
+    }
+
+    void update(FunctionContext* ctx, const Column** columns, AggDataPtr __restrict state,
+                size_t row_num) const override {
+        T column1_value;
+        if (columns[1]->is_nullable()) {
+            if (columns[1]->is_null(row_num)) {
+                return;
+            }
```

Comment:
what if all inputs are null? it seems need to count the `NULL`, rather than just skip it.

Reply:
if all inputs are null, it should return null. so skip it here is ok. i had tested this situation